### ISSUE TITLE
Threads limitation - 32?

### ISSUE THREAD

**Issue by mkrystian** [VROOM-Project/vroom#489]
Hi,

I've started the VROOM docker container in VM with 96 cores.
I set the parameter "threads" to 96 and send the huge request.

Unfortunately, the number of running threads is only 32.
I have no CPU limits set on docker.

Do you know if there is any additional limitation hardcoded in VROOM?

**config.yml**
`cliArgs:
  geometry: false # retrieve geometry (-g)
  threads: 96 # number of threads to use (-t)
  explore: 5 # exploration level to use (0..5) (-x)
  limit: '10mb' # max request size  (default was 1mb)
  logdir: '/..' # the path for the logs relative to ./src
  maxlocations: 30000 # max number of jobs/shipments locations (deafult was 1000)
  maxvehicles: 200 # max number of vehicles
  override: true # allow cli options override (-g, -t and -x)
  path: '' # VROOM path (if not in $PATH)
  port: 3000 # expressjs port
  router: 'ors' # routing backend (osrm, libosrm or ors) (default was osrm)
  timeout: 180000000 # milli-seconds (default was 300000)
  baseurl: '/' #base url for api
routingServers:
  osrm:
    car:
      host: '0.0.0.0'
      port: '5000'
    bike:
      host: '0.0.0.0'
      port: '5000'
    foot:
      host: '0.0.0.0'
      port: '5000'
  ors:
    driving-car:
      host: 'salesmen'
      port: '8082'
    driving-hgv:
      host: '0.0.0.0'
      port: '8082'
    cycling-regular:
      host: '0.0.0.0'
      port: '8082'
    cycling-mountain:
      host: '0.0.0.0'
      port: '8082'
    cycling-road:
      host: '0.0.0.0'
      port: '8082'
    cycling-electric:
      host: '0.0.0.0'
      port: '8082'
    foot-walking:
      host: '0.0.0.0'
      port: '8082'
    foot-hiking:
      host: '0.0.0.0'
      port: '8082'
`

![Screenshot 2021-04-14 at 17 10 11](https://user-images.githubusercontent.com/7273128/114736594-8172fb80-9d46-11eb-9855-58e0bc90de65.png)

**Comment by krypt-n** [comment]
As far as I understand the code, this seems to indeed be a limitation of vroom.
vroom currently uses the available threads by distributing 32 searches with different heuristic configurations onto them.
This means that if there are more than 32 threads, there are some that vroom doesn't assign any work to.

**Comment by jcoupey** [comment]
> vroom currently uses the available threads by distributing 32 searches with different heuristic configurations onto them

Exactly, so currently there is no point in using a machine with more that 32 threads, except of course to improve concurrency across several `vroom` calls.

Using more threads with the current code would mean using more heuristic parameters, that is enlarging the exploration scope, not reducing computing time. This may lead to better solutions in some cases but 1. the current tuning is already a good "all-round" setup and 2. picking different/more choices in a relevant way is not a trivial task.

On the other hand, parallelizing parts of the local search itself could be a way to use several threads within a single optimization run. I've been considering this idea for a while but did not get a chance to give it a try yet.

**Comment by dooley** [comment]
![vroom_t24](https://user-images.githubusercontent.com/1427289/120779400-54adba00-c527-11eb-86f6-b1df2ea590ba.png)

Running 1.10.0 with libosrm, i wonder why only 16 threads working? Is there any reason? Should vroom not use at least 23 threads, when 24 are available, as seen in the screenshot.

**Comment by krypt-n** [comment]
`-x 3` restricts the number of threads that vroom is able to use to 16. `-x 4` should increase this to 24 and `-x 5` to 32.

**Comment by dooley** [comment]
Ah. Thank you @krypt-n.
